CBCOMM-2503 Indexes support

danielrs opened the pull request:
Adds support for unique / nonunique indexes to collections. Also added some types and functions (`indexes.go`) that will come in handy when adding indexes to other schemas (users, edges, etc).

Make sure to also merge:
https://github.com/ClearBlade/cb-cli/pull/20

clarkbynum commented on indexes.go:
```diff
@@ -0,0 +1,83 @@
+// DiffIndexesFull takes two indexes, and returns two slices. The first slice
+// contains the items that are in `left` but not in `right` (added), and the second
+// slice returns the items that are in `right` but not in `left` (removed).
+func DiffIndexesFull(left, right []*rt.Index) ([]*rt.Index, []*rt.Index) {
```
what do you think about returning a struct that contains the key added and removed? that way users don't have to remember which variable is which

clarkbynum commented on push.go:
```diff
@@ -2006,12 +2061,17 @@ func updateCollection(meta *System_meta, collection map[string]interface{}, clie
-	// 3) if update fails, we assume the item doesn't exist, so we create the item
+	// 1) diff and update the collection indexes
+	// 3) attempt to update all of our items
+	// 4) if update fails, we assume the item doesn't exist, so we create the item
```
total nitpick but the steps are numbered wrong

danielrs commented:
Updated the `Differ` interface to take care of handling the added and dropped items using the `keep` and `drop` functions respectively. Updated old code to stop using `ListDiff` and use the new `UnsafeDiff` that uses the `Differ` interface. It's called unsafe since it deals with `interface{}` rather than concrete types.
